content/doc/book/pipeline/development.adoc

```diff
@@ -0,0 +1,145 @@
+ssh -p 8675 localhost declarative-linter < ./Jenkinsfile
+Errors encountered validating Jenkinsfile:
+WorkflowScript: 25: Not a valid section definition: 'agent'. Some extra configuration is required. @ line 25, column 1.
+   pipeline &#125;
+   ^
+
+WorkflowScript: 25: Missing required section 'agent' @ line 25, column 1.
+   pipeline &#125;
+   ^
+----
+
+.Valid Jenkinsfile
+[source,bash]
+----
+ssh -p 8675 localhost declarative-linter < ./Jenkinsfile
+Jenkinsfile successfully validated.
+----
+
+[[replay]]
+== "Replay" Pipeline Runs with Modifications
+
+Creating Pipelines and debugging failures in Pipelines can be difficult and time consuming.
+The standard development process for Pipeline changes -
+change, commit, push, run, and analyze - add quite a bit of overhead between iterations.
+Sometimes we just want to try out a quick change to a Pipeline and see the result.
+
+The "Replay" feature lets us quickly test modifications to a Pipeline
+without creating a new commit or following regular code creation process.
+To use the "Replay" feature:
+
+. Select a previously completed run in the build history.
+image::pipeline/replay-previous-run.png[Previous Pipeline Run]
+. Click "Replay" in the left menu
+image::pipeline/replay-left-bar.png[Replay Left-menu Button]
+. Make modifications and click "Run".
+In this example, we changed "ruby-2.3" to "ruby-2.4".
+image::pipeline/replay-modified.png[Replay Left-menu Button]
+. Check the results of changes
+
+Replay can be called multiple times on the same run,
+allowing for easy parallel testing of different changes.
+Replay can also be called on Pipeline runs that are still in-progress.
+As long as a Pipeline contained syntactically correct Groovy and able to start,
+it can be Replayed.
+
+[IMPORTANT]
+====
+When using Replay for larger modifications save your changes
+to a file or editor outside of Jenkins before running them.
+Syntax errors will cause a Pipeline run to not allow that run to be replayed,
+meaning your changes will be lost.
+See link:https://issues.jenkins-ci.org/browse/JENKINS-37589[JENKINS-37589]
+====
+
+[WARNING]
+====
+A Pipeline run launched using Replay may behave slightly differently than
+a Pipeline run started from an SCM detected commit or Build call.
+For Pipelines that are not part of a Multi-branch Pipeline,
+the commit information may differ for the original run and the Replayed run.
+See link:https://issues.jenkins-ci.org/browse/JENKINS-36453[JENKINS-36453]
+====
+
+== Built-in Documentation and Snippet Generator
+Do not overlook the
+<<getting-started#built-in-documentation, built-in documentation>> and the
+<<getting-started#snippet-generator, Snippet Generator>> when developing Pipelines.
+The behavior of both of these tools is based your currently installed version of
+Jenkins and related plugins.  They provide information that is accurate for your
+Jenkins instance.
+
+[[unit-test]]
+== Pipeline Unit Testing Framework
```

Comment:
I think this section should contain a NOTE admonition that this is not a tool supported as part of Jenkins Pipeline but rather is maintained by a third party.